### PR TITLE
moonraker: init at unstable-2021-07-18, nixos/moonraker: init

### PR DESCRIPTION
**Pull request by zhaofengli**
###### Motivation for this change

[Moonraker](https://github.com/Arksine/moonraker) is an HTTP API server for Klipper, and [fluidd](https://github.com/cadriel/fluidd) is a web-based client that uses the API.

#130616 is required to be able to configure the Klipper module to work with Moonraker.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/moonraker/default.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    repo = "moonraker";
+    rev = "42f61ceafa90fcfea8bffbe968e26a6fd8b61af6";
+    sha256 = "1w6l9pgs4n4nnk3h40y346bf6j3v4j4h1qnhj5dwlbwdxiqpd9gs";
+  };
+
+  installPhase = ''
+    mkdir -p $out $out/bin $out/lib
+    cp -r moonraker $out/lib
+
+    echo -e "#!/bin/sh\nexec ${pythonEnv}/bin/python $out/lib/moonraker/moonraker.py \"\$@\"" > $out/bin/moonraker
```
Comment (lovesegfault) [inline pkgs/servers/moonraker/default.nix:30]: Why don't you just use `patchShebangs`?

File: nixos/modules/misc/ids.nix [hunk 2]
```diff
@@ -347,6 +347,7 @@ in
       #mailman = 316;  # removed 2019-08-30
       zigbee2mqtt = 317;
       # shadow = 318; # unused
+      moonraker = 319;
```
Comment (lovesegfault) [inline nixos/modules/misc/ids.nix:350]: Why is DynamicUser not enough?

File: nixos/modules/services/misc/moonraker.nix [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        type = types.package;
+        default = pkgs.moonraker;
+        description = "The Moonraker package.";
+      };
+
+      klipperSocket = mkOption {
+        type = types.path;
+        example = "/run/klipper/api";
+        description = "Path to Klipper's API socket.";
+      };
```
Comment (lovesegfault) [inline nixos/modules/services/misc/moonraker.nix:24]: Set a default

File: nixos/modules/services/misc/moonraker.nix [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      group = mkOption {
+        type = types.str;
+        default = "moonraker";
+        description = "Group account under which Moonraker runs.";
+      };
+
+      listenHost = mkOption {
+        type = types.str;
+        default = "0.0.0.0";
+        example = "127.0.0.1";
```
Comment (lovesegfault) [inline nixos/modules/services/misc/moonraker.nix:57]: Make the default be the less surprising `127.0.0.1`, only expand the surface of the service if the user opts-in (by setting to `0.0.0.0`.

File: nixos/modules/services/misc/moonraker.nix [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        description = "User account under which Moonraker runs.";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "moonraker";
+        description = "Group account under which Moonraker runs.";
+      };
+
+      listenHost = mkOption {
```
Comment (lovesegfault) [inline nixos/modules/services/misc/moonraker.nix:54]: This name strikes me as odd, this is usually just called `address` for other services.

File: nixos/modules/services/misc/moonraker.nix [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        type = types.str;
+        default = "0.0.0.0";
+        example = "127.0.0.1";
+        description = "The IP or host to listen on.";
+      };
+
+      listenPort = mkOption {
```
Comment (lovesegfault) [inline nixos/modules/services/misc/moonraker.nix:61]: Likewise, this should just be called `port`.

File: nixos/modules/services/misc/moonraker.nix [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+in {
+  options = {
+    services.moonraker = {
+      enable = mkEnableOption "Moonraker, an API web server for Klipper";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.moonraker;
+        description = "The Moonraker package.";
+      };
```
Comment (lovesegfault) [inline nixos/modules/services/misc/moonraker.nix:18]: The package setting is mostly for places where there can realistically be multiple co-existing packages in `nixpkgs` that serve the same function (e.g. multiple versions of samba).

I don't think this is the case for Moonraker, we only have one package for it. If users want they can still override it with an overlay.

File: pkgs/applications/misc/fluidd/default.nix [hunk 8]
```diff
@@ -0,0 +1,33 @@
+{ lib, stdenvNoCC, fetchurl, unzip }:
+
+stdenvNoCC.mkDerivation rec {
```
Comment (lovesegfault) [inline pkgs/applications/misc/fluidd/default.nix:3]: I'm not sure, but you might want to set `dontBuild` and `dontConfigure`

**Review by lovesegfault — CHANGES_REQUESTED** [review]
Did a quick pass, are you planning on adding a module for fluidd, or instructions on setting it up with moonraker somewhere?

File: pkgs/applications/misc/fluidd/default.nix [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  unpackPhase = ''
+    mkdir www
+    unzip $src -d www
+  '';
+
+  # We use a sub-directory to accomodate other types of files that might
+  # be added in the future.
+  installPhase = ''
+    mkdir $out
+    cp -r www $out/
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/fluidd/default.nix:24]: We can't have $out/www.

File: pkgs/development/python-modules/streaming-form-data/default.nix [hunk 10]
```diff
@@ -0,0 +1,35 @@
+{ lib, fetchFromGitHub, buildPythonPackage, isPy3k,
+cython, mypy, numpy, pytest, requests-toolbelt }:
+
+buildPythonPackage rec {
+  pname = "streaming-form-data";
+  version = "1.8.1";
+  disabled = !isPy3k;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/streaming-form-data/default.nix:7]: Please use pythonOlder.

File: pkgs/development/python-modules/streaming-form-data/default.nix [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  };
+
+  nativeBuildInputs = [ cython ];
+
+  propagatedBuildInputs = [ requests-toolbelt ];
+
+  checkInputs = [ mypy numpy pytest ];
+
+  checkPhase = ''
+    make lint-mypy
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/streaming-form-data/default.nix:23]:
```suggestion
```
We don't need to run this.

File: pkgs/development/python-modules/streaming-form-data/default.nix [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    repo = "streaming-form-data";
+    rev = "v${version}";
+    sha256 = "1wnak8gwkc42ihgf0g9r7r858hxbqav2xdgqa8azid8v2ff6iq4d";
+  };
+
+  nativeBuildInputs = [ cython ];
+
+  propagatedBuildInputs = [ requests-toolbelt ];
+
+  checkInputs = [ mypy numpy pytest ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/streaming-form-data/default.nix:20]: Can we use pytestCheckHook?

Comment (zhaofengli) [reply pkgs/development/python-modules/streaming-form-data/default.nix:20]: pytestCheckHook doesn't seem to be able to use the correct path to the library, leading to module not found errors.

File: nixos/modules/services/misc/moonraker.nix (hunk 3 above)
Comment (zhaofengli) [reply nixos/modules/services/misc/moonraker.nix:24]: Ok, will add a default to `services.klipper` as well.

File: nixos/modules/misc/ids.nix (hunk 2 above)
Comment (zhaofengli) [reply nixos/modules/misc/ids.nix:350]: Moonraker keeps its own database and also requires read-write privileges to a persistent directory.

File: pkgs/servers/moonraker/default.nix (hunk 1 above)
Comment (zhaofengli) [reply pkgs/servers/moonraker/default.nix:30]: Hmm, I'm not sure how to then add it to `bin` without using yet another wrapper. This seems simpler than doing two steps.

**Comment by zhaofengli** [comment]
> Did a quick pass, are you planning on adding a module for fluidd, or instructions on setting it up with moonraker somewhere?

I'm not sure if there's any value in a NixOS module for fluidd, since it's a static webapp that can be set as the value for `locations."/".root`. People may want to set other configurations in the virtualHost (`useACMEHost`, etc.), and keeping one single line of the configuration in a separate module might confuse them. It can also be used on https://app.fluidd.xyz if the user does not want to self-host.

That being said, some instructions on using it would be nice. It will probably be somewhere in the wiki.

File: pkgs/applications/misc/fluidd/default.nix (hunk 9 above)
Comment (zhaofengli) [reply pkgs/applications/misc/fluidd/default.nix]: Ok, moved it to `$out/share/fluidd/htdocs`.

**Comment by zhaofengli** [comment]
Marking as draft again since Moonraker's self-update mechanism is peskier than it seems. Some upstream changes may be required.

File: pkgs/applications/misc/fluidd/default.nix [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+  src = fetchurl {
+    name = "fluidd-v${version}.zip";
+    url = "https://github.com/cadriel/fluidd/releases/download/v${version}/fluidd.zip";
+    sha256 = "1qwj25xvvxvm1fxx216nn2gp7js4d682mm3l4s7ns90fc5ygvc8i";
+  };
+
+  nativeBuildInputs = [ unzip ];
+
+  phases = [ "unpackPhase" "installPhase" ];
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/fluidd/default.nix:15]: Overwriting phases is not allowed.

Comment (zhaofengli) [reply pkgs/applications/misc/fluidd/default.nix:15]: Hmm, I thought it's a common pattern used for simple install-only derivations. I'll remove that.

**Comment by zhaofengli** [comment]
Been using this for a week, and it's been working well enough for me. Marking this as ready again.

**Comment by lovesegfault** [comment]
If you tackle the conflicts and ping me I'll merge it :)

**Comment by zhaofengli** [comment]
@lovesegfault Cool, just rebased and fixed the conflicts.

**Review by lovesegfault — APPROVED** [review]

**Pull request merged**